### PR TITLE
core: handle dmabuf for ofi_mr functions

### PR DESCRIPTION
**Pull request by shijin-aws**
When mr is registered with FI_MR_DMABUF bit in the flag,
    ofi_mr_attr_update should pass the fi_mr_dmabuf object
    from the attr user passed.

When inserting the mr to mr map, libfabric
    needs to translate the dmabuf object to mr_iov.

When constructing ofi_mr_info from mr_attr for ofi_mr_cache_search.
Convert dmabuf to iov if the flags has FI_MR_DMABUF.

### CODE REVIEW

**Comment by j-xiong** [comment]
Intel CI error is probably transient, will retry.
```
12:08:33  Unexpected output in rqfreeb: APPLICATION TIMED OUT, TIMEOUT = 180s
12:08:33  Program rqfreeb exited without No Errors
```

**Comment by sunkuamzn** [comment]
What about `ofi_mr_cache*` functions?

**Comment by shijin-aws** [comment]
>What about ofi_mr_cache* functions?

mr cache is currently skipped for dmabuf reg

**Comment by shijin-aws** [comment]
@j-xiong do you think we should a similar convert from dmabuf to iov for `ofi_mr_cache` path now?

**Comment by j-xiong** [comment]
@shijin-aws Yes, I think it makes sense to support this in mr cache for the case that virtual address mapping is available (i.e. base_addr != 0).

**Comment by shijin-aws** [comment]
>@shijin-aws Yes, I think it makes sense to support this in mr cache for the case that virtual address mapping is available (i.e. base_addr != 0).

Will update, thanks

**Comment by shijin-aws** [comment]
@j-xiong please see the updated commits. I didn't want to check whether `base_addr != 0` inside the mr_cache functions but think it should provider's responsibility to do that before using mr cache for dmabuf reg.

File: include/ofi_mr.h [hunk 1]
```diff
@@ -427,9 +438,9 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache,
  *				with the cache.
  */
 struct ofi_mr_entry *ofi_mr_cache_find(struct ofi_mr_cache *cache,
-				       const struct fi_mr_attr *attr);
+				       const struct fi_mr_attr *attr, uint64_t flags);
```
Comment (j-xiong) [inline include/ofi_mr.h:441]: Line is too long. Break after the comma.

File: prov/util/src/util_mr_map.c [hunk 2]
```diff
@@ -247,9 +247,13 @@ static struct fi_ops ofi_mr_fi_ops = {
 
 void ofi_mr_update_attr(uint32_t user_version, uint64_t caps,
 			const struct fi_mr_attr *user_attr,
-			struct fi_mr_attr *cur_abi_attr)
+			struct fi_mr_attr *cur_abi_attr,
+			uint64_t flags)
 {
-	cur_abi_attr->mr_iov = (struct iovec *) user_attr->mr_iov;
+	if (FI_VERSION_GE(user_version, FI_VERSION(1, 20)) && (flags & FI_MR_DMABUF))
```
Comment (j-xiong) [inline prov/util/src/util_mr_map.c:253]: Line too long.

File: include/ofi_mr.h [hunk 3]
```diff
@@ -296,6 +297,19 @@ static inline bool ofi_mr_all_host(struct ofi_mr **mr, size_t count)
...
 
+static inline
+void ofi_mr_get_iov_from_dmabuf(struct iovec *iov,
+				 const struct fi_mr_dmabuf *dmabuf,
+				 size_t count)
+{
+	int i;
+
+	for (i = 0; i < count; i++) {
+		iov[i].iov_base = (void *) ((uintptr_t) dmabuf[i].base_addr + dmabuf[i].offset);
```
Comment (j-xiong) [inline include/ofi_mr.h:308]: Line too long.

File: prov/efa/src/efa_mr.c [hunk 4]
```diff
@@ -634,7 +634,7 @@ int efa_mr_is_cuda_memory_freed(struct efa_mr *efa_mr, bool *freed)
  * 		negative libfabric error code on failure.
  */
 static
-int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr)
+int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr, uint64_t flags)
```
Comment (j-xiong) [inline prov/efa/src/efa_mr.c:637]: Line too long.

File: prov/efa/src/efa_mr.c [hunk 5]
```diff
@@ -750,14 +750,14 @@ int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_att
 }
 #else /* HAVE_CUDA */
 static
-int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr)
+int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr, uint64_t flags)
```
Comment (j-xiong) [inline prov/efa/src/efa_mr.c:753]: Line too long.

**Review by j-xiong — COMMENTED** [review]
Mostly fine. Some nitpick about the line length.

File: prov/efa/src/efa_mr.c (hunk 4 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: fixed

File: prov/efa/src/efa_mr.c (hunk 5 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: fixed

File: include/ofi_mr.h (hunk 3 above)
Comment (shijin-aws) [reply include/ofi_mr.h]: fixed

File: prov/util/src/util_mr_map.c (hunk 2 above)
Comment (shijin-aws) [reply prov/util/src/util_mr_map.c]: fixed

File: include/ofi_mr.h (hunk 1 above)
Comment (shijin-aws) [reply include/ofi_mr.h]: fixed

**Comment by j-xiong** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
@j-xiong ok to merge?

**Review by j-xiong — APPROVED** [review]

**Pull request merged**